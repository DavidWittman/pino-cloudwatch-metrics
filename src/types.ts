import { Logger } from 'pino'

export enum Unit {
  None = 'None',
  Seconds = 'Seconds',
  Microseconds = 'Microseconds',
  Milliseconds = 'Milliseconds',
  Bytes = 'Bytes',
  Kilobytes = 'Kilobytes',
  Megabytes = 'Megabytes',
  Gigabytes = 'Gigabytes',
  Terabytes = 'Terabytes',
  Bits = 'Bits',
  Kilobits = 'Kilobits',
  Megabits = 'Megabits',
  Gigabits = 'Gigabits',
  Terabits = 'Terabits',
  Percent = 'Percent',
  Count = 'Count',
  BytesPerSecond = 'Bytes/Second',
  KilobytesPerSecond = 'Kilobytes/Second',
  MegabytesPerSecond = 'Megabytes/Second',
  GigabytesPerSecond = 'Gigabytes/Second',
  TerabytesPerSecond = 'Terabytes/Second',
  BitsPerSecond = 'Bits/Second',
  KilobitsPerSecond = 'Kilobits/Second',
  MegabitsPerSecond = 'Megabits/Second',
  GigabitsPerSecond = 'Gigabits/Second',
  TerabitsPerSecond = 'Terabits/Second',
  CountPerSecond = 'Count/Second',
}

export interface MetricDefinition {
  Name: string
  Unit: Unit
}

export interface CloudWatchMetric {
  Namespace: string
  Dimensions: string[][]
  Metrics: MetricDefinition[]
}

export interface EmbeddedMetricFormat {
  _aws: {
    Timestamp: number
    CloudWatchMetrics: CloudWatchMetric[]
  }
}

export interface MetricValue {
  value: number
  unit?: Unit
}

export type Metrics = Record<string, number | MetricValue>
export type Dimensions = Record<string, string | number>

export interface MetricLogger extends Logger {
  metric(metrics: Record<string, MetricValue | number>): MetricBuilder
}

export interface MetricBuilder extends Logger {
  dimensions(dimensions: Dimensions): this
  namespace(namespace: string): this
}
