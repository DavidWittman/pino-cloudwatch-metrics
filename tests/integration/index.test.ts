import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
import { pinoCloudwatchMetrics } from '../../src/index.js'
import { Unit } from '../../src/types.js'
import pino from 'pino'
import path from 'path'
import os from 'os'
import fs from 'fs'

describe('pinoCloudwatchMetrics', () => {
  let baseLogger: pino.Logger
  let tmpfile: string

  beforeEach(() => {
    tmpfile = path.join(
      os.tmpdir(),
      `pino-cloudwatch-metrics-${Date.now()}.log`,
    )
    baseLogger = pino(
      { base: null },
      pino.destination({
        dest: tmpfile,
        sync: true,
      }),
    )
    // Mock system time for consistent snapshots
    vi.useFakeTimers()
    vi.setSystemTime(new Date('2025-01-01T00:00:00.000Z'))
  })

  afterEach(() => {
    vi.useRealTimers()
  })

  it('should log a metric with no dimensions', async () => {
    const logger = pinoCloudwatchMetrics()(baseLogger)

    logger
      .metric({
        Latency: { value: 200, unit: Unit.Milliseconds },
        RequestCount: 5,
      })
      .info('Test log with no dimensions in default namespace')

    const output = await fs.promises.readFile(tmpfile, 'utf-8')
    expect(output).toMatchSnapshot()
  })

  it('should log a metric in a custom namespace', async () => {
    const logger = pinoCloudwatchMetrics({ defaultNamespace: 'MilkService' })(
      baseLogger,
    )

    logger
      .metric({
        Latency: { value: 4, unit: Unit.Seconds },
        FileSize: { value: 42, unit: Unit.Megabytes },
      })
      .error('Test error log in custom namespace')

    const output = await fs.promises.readFile(tmpfile, 'utf-8')
    expect(output).toMatchSnapshot()
  })

  it('should log a metric with multiple dimensions in a custom namespace using .namespace()', async () => {
    const logger = pinoCloudwatchMetrics()(baseLogger)

    logger
      .metric({
        Latency: { value: 200, unit: Unit.Milliseconds },
        RequestCount: 5,
      })
      .dimensions({
        ServiceName: 'MilkService',
        Region: 'us-east-1',
      })
      .namespace('MilkService/Metrics')
      .info(
        { requestId: 'a9b2b1ea-0e9b-4e7e-8827-282d6d6ba659' },
        'Test log with dimensions',
      )

    const output = await fs.promises.readFile(tmpfile, 'utf-8')
    expect(output).toMatchSnapshot()
  })
})
