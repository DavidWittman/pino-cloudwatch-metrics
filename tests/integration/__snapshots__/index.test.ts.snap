// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pinoCloudwatchMetrics > should log a metric in a custom namespace 1`] = `
"{"level":50,"time":1735689600000,"_aws":{"Timestamp":1735689600000,"CloudWatchMetrics":[{"Namespace":"MilkService","Dimensions":[[]],"Metrics":[{"Name":"Latency","Unit":"Seconds"},{"Name":"FileSize","Unit":"Megabytes"}]}]},"Latency":4,"FileSize":42,"msg":"Test error log in custom namespace"}
"
`;

exports[`pinoCloudwatchMetrics > should log a metric with multiple dimensions in a custom namespace using .namespace() 1`] = `
"{"level":30,"time":1735689600000,"_aws":{"Timestamp":1735689600000,"CloudWatchMetrics":[{"Namespace":"MilkService/Metrics","Dimensions":[["ServiceName","Region","SomeIdentifier"]],"Metrics":[{"Name":"Latency","Unit":"Milliseconds"},{"Name":"RequestCount","Unit":"None"}]}]},"ServiceName":"MilkService","Region":"us-east-1","SomeIdentifier":"123","Latency":200,"RequestCount":5,"requestId":"a9b2b1ea-0e9b-4e7e-8827-282d6d6ba659","msg":"Test log with dimensions"}
"
`;

exports[`pinoCloudwatchMetrics > should log a metric with no dimensions 1`] = `
"{"level":30,"time":1735689600000,"_aws":{"Timestamp":1735689600000,"CloudWatchMetrics":[{"Namespace":"Pino","Dimensions":[[]],"Metrics":[{"Name":"Latency","Unit":"Milliseconds"},{"Name":"RequestCount","Unit":"None"}]}]},"Latency":200,"RequestCount":5,"msg":"Test log with no dimensions in default namespace"}
"
`;
